addon-test-support/@ember/test-helpers/dom/fire-event.js

```diff
@@ -0,0 +1,225 @@
+import { merge } from '@ember/polyfills';
+
+const DEFAULT_EVENT_OPTIONS = { bubbles: true, cancelable: true };
+export const KEYBOARD_EVENT_TYPES = Object.freeze(['keydown', 'keypress', 'keyup']);
+const MOUSE_EVENT_TYPES = [
+  'click',
+  'mousedown',
+  'mouseup',
+  'dblclick',
+  'mouseenter',
+  'mouseleave',
+  'mousemove',
+  'mouseout',
+  'mouseover',
+];
+const FILE_SELECTION_EVENT_TYPES = ['change'];
+
+/**
+  @method fireEvent
+  @param {HTMLElement} element
+  @param {String} type
+  @param {Object} [options]
+
+  @private
+*/
+export default function fireEvent(element, type, options = {}) {
+  if (!element) {
+    throw new Error('Must pass an element to `fireEvent`');
+  }
+
+  let event;
+  if (KEYBOARD_EVENT_TYPES.indexOf(type) > -1) {
+    event = buildKeyboardEvent(type, options);
+  } else if (MOUSE_EVENT_TYPES.indexOf(type) > -1) {
+    let rect;
+    if (element instanceof Window) {
+      rect = element.document.documentElement.getBoundingClientRect();
+    } else if (element instanceof Document) {
+      rect = element.documentElement.getBoundingClientRect();
+    } else if (element instanceof HTMLElement || element instanceof SVGElement) {
+      rect = element.getBoundingClientRect();
+    } else {
+      return;
+    }
+
+    let x = rect.left + 1;
+    let y = rect.top + 1;
+    let simulatedCoordinates = {
+      screenX: x + 5, // Those numbers don't really mean anything.
+      screenY: y + 95, // They're just to make the screenX/Y be different of clientX/Y..
+      clientX: x,
+      clientY: y,
+    };
+
+    event = buildMouseEvent(type, merge(simulatedCoordinates, options));
+  } else if (FILE_SELECTION_EVENT_TYPES.indexOf(type) > -1 && element.files) {
+    event = buildFileEvent(type, element, options);
+  } else {
+    event = buildBasicEvent(type, options);
+  }
+
+  element.dispatchEvent(event);
+}
+
+/**
+  @method buildBasicEvent
+  @param {String} type
+  @param {Object} [options]
+  @return {Event}
+  @private
+*/
+function buildBasicEvent(type, options = {}) {
+  let event = document.createEvent('Events');
+
+  let bubbles = options.bubbles !== undefined ? options.bubbles : true;
+  let cancelable = options.cancelable !== undefined ? options.cancelable : true;
+
+  delete options.bubbles;
+  delete options.cancelable;
+
+  // bubbles and cancelable are readonly, so they can be
+  // set when initializing event
+  event.initEvent(type, bubbles, cancelable);
+  merge(event, options);
+  return event;
+}
+
+/**
+  @method buildMouseEvent
+  @param {String} type
+  @param {Object} [options]
+  @return {Event}
+  @private
+*/
+function buildMouseEvent(type, options = {}) {
+  let event;
+  try {
+    event = document.createEvent('MouseEvents');
+    let eventOpts = merge(merge({}, DEFAULT_EVENT_OPTIONS), options);
+    event.initMouseEvent(
```

Comment:
IDK if this targets Ember 3.0, but if it does, then we don't need this as the `MouseEvent` constructor works in IE11+

Reply:
I don’t plan to change browser support as part of this effort, however we should add a comment to indicate it can be removed if IE8/IE9 are not supported.

Reply:
9 and 10. Probably the same will happen to `KeyboardEvent`

Reply:
TODO: Add support for pointer events? That might unify tap/click.